### ISSUE TITLE
Changing block size after negotiation

### ISSUE THREAD

**Issue by sabor12345** [obgm/libcoap#1290]
## Environment

- Build System:             [Make]
- Operating System:         [Linux]
- Operating System Version: [ Debian GNU/Linux 12]
- Hosted Environment:       [None)]

## libcoap Configuration Summary

`$ ./configure --prefix=/home/libcoap_2023_10_22/libcoap/build --disable-manpages
#define PACKAGE_NAME "libcoap"
| #define PACKAGE_TARNAME "libcoap"
| #define PACKAGE_VERSION "4.3.1"
| #define PACKAGE_STRING "libcoap 4.3.1"

## Problem Description

It is possible to change the block size after negotiation. Also when sending a request with a new block size the server is sending the wrong block number. 

### Expected Behavior

According to the RFC7959:
"Any further block-wise requests for blocks beyond the first one MUST indicate the same block size that was used by the server in the response for the first request that gave a desired size using a Block2 Option."

Only my thoughts:

I guess if any further request indicates a different block size, the server should detect it and not allow it?
Also, the client should not accept responses with a different block size after block negotiation if the M bit is not set? 

### Actual Behavior

It is possible to send a request with a different block size to the server after block size negotiation. The server proceeds with this new block size. In this particular case, the server also returns the wrong block number, see below.

### Steps to reproduce

Using the server and client examples in the example folder.

## Debug Logs
client: ./coap-client -v 7 -m get -b 64 coap://[::1]/.well-known/core
server: ./coap-server -v 7
```
client sends this: 
v:1 t:CON c:GET i:f99d {01} [ Uri-Path:.well-known, Uri-Path:core, Block2:0/_/64 ]

receives this: v:1 t:ACK c:2.05 i:f99d {01} [ ETag:0x01, Content-Format:application/link-format, Block2:0/M/64, Size2:151 ] :: '</>;title="General Info";ct=0,</time>;if="clock";rt="ticks";titl'

The client sends this with a different block size:
v:1 t:CON c:GET i:f99e {02000000000002} [ Uri-Path:.well-known, Uri-Path:core, Block2:1/_/16 ]

server produces this:

Nov 28 13:38:09.749 DEBG found Block option, block size is 16, block nr. 1
Nov 28 13:38:09.749 DEBG new Block size is 16, block number 3 completed

v:1 t:ACK c:2.05 i:f99e {02000000000002} [ ETag:0x01, Content-Format:application/link-format, Block2:1/M/16, Size2:151 ] :: 'al Info";ct=0,</'

```
It prints that block number 3 is completed but it still sends block number 1 in the response. The client and server will continue with this new block size until the transfer is completed.

**Comment by mrdeep1** [comment]
@sabor12345 Fixed in #1294, which is waiting to be merged.  Please confirm that this works for you.

**Comment by mrdeep1** [comment]
Can this be closed now?
